frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -1304,16 +1314,112 @@ typeConstructorInitialQuery(Context* context, const Type* t)
+
+  auto parentMod = parsing::idToParentModule(context, id);
+  auto modName = "chpl__internal_" + parentMod.symbolName(context).str() + "_" + name.str();
+  auto builder = Builder::createForGeneratedCode(context, modName.c_str(), parentMod.symbolPath());
+  auto dummyLoc = parsing::locateId(context, id);
+
+  // Add a top-level 'use' of the module containing the type being constructed.
+  // This exists to support fields that reference other types in that module.
+  {
+    auto useName = parentMod.symbolName(context);
+    auto ident = Identifier::build(builder.get(), dummyLoc, useName);
+    auto vis = VisibilityClause::build(builder.get(), dummyLoc, std::move(ident));
+
+    AstList alist;
+    alist.push_back(std::move(vis));
+    auto useStmt = Use::build(builder.get(), dummyLoc,
+                              Decl::DEFAULT_VISIBILITY, std::move(alist));
+    builder->addToplevelExpression(std::move(useStmt));
+  }
+
+  // Build formals of type constructor
+  AstList formalAst;
+  for(auto& arg : fieldAsts) {
+    auto var = arg->toVariable();
+    owned<AttributeGroup> attr;
+    if (var->attributeGroup()) {
+      auto attrCopy = var->attributeGroup()->copy();
+      AttributeGroup* asdf = attrCopy.release()->toAttributeGroup();
+      attr = std::make_unique<AttributeGroup>(*asdf);
+    }
+    auto typeExpr = var->typeExpression();
+    auto initExpr = var->initExpression();
+    auto kind = var->kind() == Variable::PARAM ? Variable::PARAM : Variable::TYPE;
+    owned<AstNode> formal = Formal::build(builder.get(), dummyLoc,
+                                std::move(attr),
+                                var->name(),
+                                (Formal::Intent)kind,
+                                typeExpr ? typeExpr->copy() : nullptr,
+                                initExpr ? initExpr->copy() : nullptr);
+    formalAst.push_back(std::move(formal));
+  }
+
+  auto genFn = Function::build(builder.get(), dummyLoc, {},
+                               Decl::Visibility::PUBLIC,
+                               Decl::Linkage::DEFAULT_LINKAGE,
+                               {},
+                               name,
+                               false, false, Function::Kind::PROC,
+                               {},
+                               Function::ReturnIntent::DEFAULT_RETURN_INTENT,
+                               false, false, false,
+                               std::move(formalAst), {}, {}, {}, {});
+
+  builder->noteChildrenLocations(genFn.get(), dummyLoc);
+  builder->addToplevelExpression(std::move(genFn));
+
+  // Finalize the uAST and obtain the BuilderResult
+  auto res = builder->result();
+
+  // Store the BuilderResult for later, using the module's path as the
+  // query key.
+  auto modPath = res.topLevelExpression(0)->id().symbolPath();
+  parsing::setCompilerGeneratedBuilder(context, modPath, std::move(res));
+
+  //
+  // Re-acquire the BuilderResult
```

Comment:
Can all the generation code "before" re-acquiring the builder result be moved into a helper to reduce the size of this function?

Reply:
Actually, I'm inclined to include the re-acquisition of the BuilderResult so that if I/we can figure out a way to make that cleaner later we can just have the helper function return a BuilderResult. Thoughts?

Reply:
That sounds good to me 👍